src/main/scala/io/iohk/ethereum/blockchain/sync/RegularSync.scala

```diff
@@ -52,10 +58,56 @@ class RegularSync(
       askForHeaders()
   }
 
-  def running: Receive = handleCommonMessages orElse {
-    case ResumeRegularSync =>
-      askForHeaders()
+  private def resumeRegularSync(): Unit = {
+    resumeRegularSyncTimeout.foreach(_.cancel)
+    resumeRegularSyncTimeout = None
+    askForHeaders()
+  }
 
+  /**
+    * Handles broadcasted blocks, processing them and inserting them to the blockchain if necessary
+    * FIXME: Current implementation is naive in that NewBlock msgs that can't be immediately processed are discarded
+    *        and requested again to the peers
+    */
+  def handleBroadcastedBlockMessages: Receive = {
+    case MessageFromPeer(NewBlock(newBlock, _), peerId) =>
+      //we allow inclusion of new block only if we are not syncing / reorganising chain
+      if (headersQueue.isEmpty && waitingForActor.isEmpty) {
+        val currentBestBlock = appStateStorage.getBestBlockNumber()
+        val maybeBlockParentTd = blockchain.getBlockHeaderByHash(newBlock.header.parentHash)
+          .flatMap(b => blockchain.getTotalDifficultyByHash(b.hash))
+        val maybeCurrentBlockTd = blockchain.getTotalDifficultyByNumber(currentBestBlock)
+
+        (maybeBlockParentTd, maybeCurrentBlockTd) match {
+          case (_, None) =>
+            //Something went wrong, we don't have the total difficulty of the latest block
+            //TODO: Investigate if we can recover from this error [EC-165]
+            throw new IllegalStateException(s"No total difficulty for the latest block with number $currentBestBlock")
```

Comment:
throw here will just restart the actor, do we need that?

Reply:
Maybe we change that the actor is not restarted in this case or just log the error? I decided to throw the exception as it was also handled this way when [receiving a block and not having it's parent total difficulty](https://github.com/input-output-hk/mantis/blob/phase/daedalus/src/main/scala/io/iohk/ethereum/blockchain/sync/RegularSync.scala#L243), as discussed in https://github.com/input-output-hk/mantis/pull/148#discussion_r112447736.

Reply:
I'd say we should just crash. Any rationale for restarting the actor?